test/test_shape.py

```diff
@@ -147,6 +148,93 @@ def _m1bm1_dim_checker(x: TensorType[-1, "b", -1]):
         _m1bm1_dim_checker(x)
 
 
+def test_str_str_dim1():
+    @typeguard.typechecked
+    def func(x: TensorType["a":"x"]):
+        pass
+
+    func(torch.ones(3))
+    func(torch.ones(2))
+    with pytest.raises(TypeError):
+        func(torch.tensor(3.0))
+    with pytest.raises(TypeError):
+        func(torch.ones(3, 3))
+
+
+def test_str_str_dim2():
+    @typeguard.typechecked
+    def func(x: TensorType["a":"x", "b":"x"]):
+        pass
+
+    func(torch.ones(3, 3))
+    func(torch.ones(2, 2))
+    with pytest.raises(TypeError):
+        func(torch.tensor(3.0))
+    with pytest.raises(TypeError):
+        func(torch.ones(3))
+    with pytest.raises(TypeError):
+        func(torch.ones(3, 2))
+    with pytest.raises(TypeError):
+        func(torch.ones(2, 3))
+
+
+def test_str_str_dim_complex():
+    @typeguard.typechecked
+    def func(x: TensorType["a":"x", "b":"x", "x", "a", "b"]) -> TensorType["c":"x"]:
+        return torch.ones(x.shape[0])
+
+    func(torch.ones(3, 3, 3, 3, 3))
+    func(torch.ones(2, 2, 2, 2, 2))
+    with pytest.raises(TypeError):
+        func(torch.ones(1, 2, 2, 2, 2))
+    with pytest.raises(TypeError):
+        func(torch.ones(2, 1, 2, 2, 2))
+    with pytest.raises(TypeError):
+        func(torch.ones(2, 2, 1, 2, 2))
+    with pytest.raises(TypeError):
+        func(torch.ones(2, 2, 2, 1, 2))
+    with pytest.raises(TypeError):
+        func(torch.ones(2, 2, 2, 2, 1))
+
+    @typeguard.typechecked
+    def func2(x: TensorType["a":"x", "b":"x", "x", "a", "b"]) -> TensorType["c":"x"]:
+        return torch.ones(x.shape[0] + 1)
+
+    with pytest.raises(TypeError):
+        func2(torch.ones(2, 2, 2, 2, 2))
+
+    @typeguard.typechecked
+    def func3(x: TensorType["a":"x", "b":"x", "x", "a", "b"]) -> TensorType["c":"x"]:
+        return torch.ones(x.shape[0], x.shape[0])
+
+    with pytest.raises(TypeError):
+        func3(torch.ones(2, 2, 2, 2, 2))
+
+
+def test_str_str_dim_fixed_num():
+    @typeguard.typechecked
+    def func(x: TensorType["a":"x"]) -> TensorType["x":3]:
+        return torch.ones(x.shape[0])
+
+    func(torch.ones(3))
+    with pytest.raises(TypeError):
+        func(torch.ones(2))
+
+
+def test_str_str_dim_fixed_names():
+    @typeguard.typechecked
+    def func(x: TensorType["a":"x", is_named]) -> TensorType["x":3]:
+        return torch.ones(x.shape[0])
+
+    func(torch.ones(3, names=["a"]))
+    with pytest.raises(TypeError):
+        func(torch.ones(3))
+    with pytest.raises(TypeError):
+        func(torch.ones(3, names=["b"]))
+    with pytest.raises(TypeError):
+        func(torch.ones(2, names=["a"]))
+
```

Comment:
A test that `names=["x"]` fails would be good to have as well, given that that's the other name.

Or perhaps we should make it permissive and accept either name on a named dimension? No strong feelings either way, just that it should be checked + documented whichever way around we do it.

Reply:
good idea!  I do think that we should strictly check that the name mates the left, though.  Test added.